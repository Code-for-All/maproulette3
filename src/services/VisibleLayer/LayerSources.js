import _isEmpty from 'lodash/isEmpty'
import _find from 'lodash/find'
import PropTypes from 'prop-types'
import messages from './Messages'

export const layerSourceShape = PropTypes.shape({
  /** Unique id for layer */
  layerId: PropTypes.string.isRequired,
  /** Human-readable name of layer formatted as react-intl message */
  name: PropTypes.object,
  /** Tile server URL. Supports substitutions: {x}, {y}, {z}, {accessToken} */
  url: PropTypes.string.isRequired,
  /** Human-readable attribution of layer formatted as react-intl message */
  attribution: PropTypes.object,
  /** Width/height of tiles */
  tileSize: PropTypes.number,
  /**
   * Set to true to adjust non-retina tiles for retina displays. tileSize may
   * also need to be adjusted.
   *
   * @see See http://leafletjs.com/reference-1.0.3.html#tilelayer-detectretina
   */
  detectRetina: PropTypes.bool,
  /** Access token for substitution in tile server url */
  accessToken: PropTypes.string,
  /** Set to true to mark as the default layer */
  default: PropTypes.bool,
  /** Set to false if layer should not be offered to users for selection */
  isSelectable: PropTypes.bool,
  /** Set to true for dynamically-created layers, such as custom basemaps */
  isDynamic: PropTypes.bool,
})

export const OPEN_STREET_MAP = 'OpenStreetMap'
export const OPEN_CYCLE_MAP = 'OpenCycleMap'
export const BING = 'Bing'
export const MAPBOX_STREETS = 'Mapbox'
export const MAPBOX_LIGHT = 'MapboxLight'
export const MAPBOX_SATELLITE_STREETS = 'MapboxSatellite'

/** Array of available layer sources */
export const LayerSources = [{
    layerId: OPEN_STREET_MAP,
    name: messages.openStreetMapName,
    url: process.env.REACT_APP_OPEN_STREET_MAP_TILESERVER_URL,
    attribution: messages.openStreetMapAttribution,
  }, {
    layerId: OPEN_CYCLE_MAP,
    name: messages.openCycleMapName,
    url: process.env.REACT_APP_OPEN_CYCLE_MAP_TILESERVER_URL,
    attribution: messages.openCycleMapAttribution,
  }, {
    layerId: BING,
    name: messages.bingName,
    url: process.env.REACT_APP_BING_MAP_TILESERVER_URL,
    attribution: messages.bingAttribution,
    detectRetina: true,
    tileSize: 256,
  }]

// If a Mapbox access/api token has been provided, then add Mapbox layer
// sources.
if (!_isEmpty(process.env.REACT_APP_MAPBOX_ACCESS_TOKEN)) {
  LayerSources.push({
    layerId: MAPBOX_STREETS,
    name: messages.mapboxStreetsName,
    url: process.env.REACT_APP_MAPBOX_STREETS_MAP_TILESERVER_URL,
    accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
    attribution: messages.mapboxAttribution,
    default: true,
  })

  LayerSources.push({
    layerId: MAPBOX_SATELLITE_STREETS,
    name: messages.mapboxSatelliteName,
    url: process.env.REACT_APP_MAPBOX_SATELLITE_STREETS_MAP_TILESERVER_URL,
    accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
    attribution: messages.mapboxAttribution,
  })

  LayerSources.push({
    layerId: MAPBOX_LIGHT,
    name: messages.mapboxLightName,
    url: process.env.REACT_APP_MAPBOX_LIGHT_MAP_TILESERVER_URL,
    accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
    attribution: messages.mapboxAttribution,
  })
}

/**
 * Returns a default layer source for use in situations where no layer source
 * has been specified.
 */
export const defaultLayerSource = function() {
  const configuredDefault = _find(LayerSources, {default: true})
  return configuredDefault ? configuredDefault : LayerSources[0]
}

/**
 * Retrieves the (static) layer source with a matching id. Dynamic layer
 * sources are not searched.
 */
export const layerSourceWithId = function(layerId) {
  return _find(LayerSources, {layerId})
}

/**
 * Create and return a dynamic layer source with the given layerId
 * and url. Primarily intended for use with custom basemaps.
 */
export const createDynamicLayerSource = function(layerId, url) {
  return {
    layerId,
    name: messages.customName,
    url,
    isSelectable: false,
    isDynamic: true,
  }
}
