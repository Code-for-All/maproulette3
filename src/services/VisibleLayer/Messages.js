import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with VisibleLayer
 */
export default defineMessages({
  openStreetMapName: {
    id: "map.name.openStreetMap",
    defaultMessage: "OpenStreetMap",
  },

  openStreetMapAttribution: {
    id: "map.attribution.openStreetMap",
    defaultMessage: "&copy; <a href='http://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='http://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
  },

  openCycleMapName: {
    id: "map.name.openCycleMap",
    defaultMessage: "OpenCycleMap",
  },

  openCycleMapAttribution: {
    id: "map.attribution.openCycleMap",
    defaultMessage: "&copy; OpenCycleMap, Map data &copy; <a href='http://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='http://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
  },

  bingName: {
    id: "map.name.bing",
    defaultMessage: "Bing",
  },

  bingAttribution: {
    id: "map.attribution.bing",
    defaultMessage: "Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community",
  },

  mapboxStreetsName: {
    id: "map.name.mapbox",
    defaultMessage: "Mapbox",
  },

  mapboxLightName: {
    id: "map.name.mapboxLight",
    defaultMessage: "Mapbox Light",
  },

  mapboxSatelliteName: {
    id: "map.name.mapboxSatellite",
    defaultMessage: "Mapbox Satellite",
  },

  mapboxAttribution: {
    id: "map.attribution.mapbox",
    defaultMessage: "&copy <a href='https://www.mapbox.com/about/maps/'>Mapbox</a>, &copy <a href='http://www.openstreetmap.org/copyright'>OpenStreetMap</a> <strong><a href='https://www.mapbox.com/map-feedback/' target='_blank'>Improve this map</a></strong>",
  },

  customName: {
    id: "map.name.custom",
    defaultMessage: "Custom",
  }
})
