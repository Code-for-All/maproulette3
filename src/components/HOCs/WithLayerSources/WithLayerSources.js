import React, { Component } from 'react';
import { LayerSources } from '../../../services/VisibleLayer/LayerSources'

export default function WithLayerSources(WrappedComponent) {
  return class extends Component {
    render() {
      return <WrappedComponent layerSources={LayerSources} {...this.props} />
    }
  }
}
