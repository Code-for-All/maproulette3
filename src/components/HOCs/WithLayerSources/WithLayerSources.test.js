import React, { Component } from 'react'
import WithLayerSources from './WithLayerSources'
import { LayerSources } from '../../../services/VisibleLayer/LayerSources'

let WrappedComponent = null

beforeEach(() => {
  WrappedComponent = WithLayerSources(() => <div className="child-control" />)
})

test("the wrapped component is rendered", () => {
  const wrapper = shallow(
    <WrappedComponent />
  )

  expect(wrapper).toMatchSnapshot()
})

test("layer sources are made available to the wrapped component", () => {
  const wrapper = shallow(
    <WrappedComponent />
  )

  expect(wrapper.props().layerSources).toEqual(LayerSources)
})
