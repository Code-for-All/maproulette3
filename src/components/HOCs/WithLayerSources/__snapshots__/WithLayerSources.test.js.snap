// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the wrapped component is rendered 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Component />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "layerSources": Array [
        Object {
          "attribution": Object {
            "defaultMessage": "&copy; <a href='https://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='https://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
            "id": "map.attribution.openStreetMap",
          },
          "layerId": "OpenStreetMap",
          "name": Object {
            "defaultMessage": "OpenStreetMap",
            "id": "map.name.openStreetMap",
          },
          "url": "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
        },
        Object {
          "accessToken": "",
          "attribution": Object {
            "defaultMessage": "&copy; OpenCycleMap, Map data &copy; <a href='https://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='https://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
            "id": "map.attribution.openCycleMap",
          },
          "layerId": "OpenCycleMap",
          "name": Object {
            "defaultMessage": "OpenCycleMap",
            "id": "map.name.openCycleMap",
          },
          "url": "https://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
        },
        Object {
          "attribution": Object {
            "defaultMessage": "Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community",
            "id": "map.attribution.bing",
          },
          "layerId": "Bing",
          "name": Object {
            "defaultMessage": "Bing",
            "id": "map.name.bing",
          },
          "url": "https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}",
        },
      ],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "layerSources": Array [
          Object {
            "attribution": Object {
              "defaultMessage": "&copy; <a href='https://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='https://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
              "id": "map.attribution.openStreetMap",
            },
            "layerId": "OpenStreetMap",
            "name": Object {
              "defaultMessage": "OpenStreetMap",
              "id": "map.name.openStreetMap",
            },
            "url": "https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png",
          },
          Object {
            "accessToken": "",
            "attribution": Object {
              "defaultMessage": "&copy; OpenCycleMap, Map data &copy; <a href='https://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='https://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
              "id": "map.attribution.openCycleMap",
            },
            "layerId": "OpenCycleMap",
            "name": Object {
              "defaultMessage": "OpenCycleMap",
              "id": "map.name.openCycleMap",
            },
            "url": "https://{s}.tile.opencyclemap.org/cycle/{z}/{x}/{y}.png",
          },
          Object {
            "attribution": Object {
              "defaultMessage": "Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community",
              "id": "map.attribution.bing",
            },
            "layerId": "Bing",
            "name": Object {
              "defaultMessage": "Bing",
              "id": "map.name.bing",
            },
            "url": "https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}",
          },
        ],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
